### PR TITLE
feat: Support Hermes (react-native) SourceMaps

### LINKED ISSUE #21: Implement hermes source maps
We currently cannot handle hermes source maps but it should be possible with some larger changes in the library (refs https://github.com/getsentry/sentry-react-native/issues/649)

We probably want to normalize `x_facebook_sources` somehow because that contains scope mapping information which is exactly what we need. This means we no longer need to access minified sources for figuring out the function name.

## Rough Overview

Hermes internally compiles javascript sources to bytecode directly which means there is actually no minified javascript behind the scenes. Currently we use the minified source to figure out a few things that are crucial for processing:

* we use the comment in the minified file to locate the source map (https://github.com/getsentry/rust-sourcemap/blob/a086b928708d098808639080f118da5990607c0f/src/detector.rs#L47-L60)
* we use the minified file to reverse tokenize from the error location to the function scope (heuristic) (https://github.com/getsentry/rust-sourcemap/blob/a086b928708d098808639080f118da5990607c0f/src/types.rs#L592-L601)

These two things are generally used by sentry and assume some behavior already. In particular we assume that the reported function name is the minified file.

With Hermes both of those assumptions are invalidated.  First of all because there is no minified file we cannot use our heuristics. Thankfully though Hermes source maps have a second mapping in them called `x_facebook_sources` which contains mappings from a token location to the enclosing function. The metro symbolicator already uses this to map back to a function name.

Secondly because there is just one huge source map for all files in a react native app we need a separate algorithm to map from the reported filename in a stack to the section in the source map. In the past (with RAM bundles) we rewrote them to virtual files. (see the [ram bundle code](https://github.com/getsentry/rust-sourcemap/blob/master/src/ram_bundle.rs)). Here we probably want to add an abstraction at a higher level where we provide a file name and it returns where in the source map it would be and which source index in it (as there are multiple virtual files in it).

This means we want to change the "get original function name" logic to make the minified source view optional for the hermes case where the source map has enough information to get back to the original name.

## Changes in Sentry

After we fixed this here we also need to do some changes in sentry obviously. Currently we rewrite source maps in sentry-cli on the way up and if they have not been rewritten then, they are rewritten again by sentry on loading. We probably do not want to do anything like this with hermes source maps.

The changes there will likely be these:

1. use the 0 based indexing for hermes instead of 1 indexed like we do for normal source maps
2. add a layer of indirection for selecting source maps for hermes / react-native to not look for the minified source file but navigate to the source map directly (assume it always comes with the same fixed name?)
3. not pass the minified file when looking for the original function name.

### PR DESCRIPTION
**Pull request by Swatinem**
fixes #21 

This exposes a new `SourceMapHermes` type, which can look up the scope name via a bytecode offset.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Yeah. This is all a bit meh. I think one way we could go forward is instead of exposing the source map as the primary API to have view objects over them that provide the most common operations.

We already effectively do something like this in symbolic where a source map often only makes sense when paired with the source view of the minified code.

Might be worth looking into the firefox developer tools. They might have found better abstractions.

File: src/hermes.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        // Find the closest mapping, just like here:
+        // https://github.com/facebook/metro/blob/63b523eb20e7bdf62018aeaf195bb5a3a1a67f36/packages/metro-symbolicate/src/SourceMetadataMapConsumer.js#L204-L231
+        let mapping =
+            function_map
+                .mappings
+                .binary_search_by(|o| match o.line.cmp(&token.get_src_line()) {
+                    Ordering::Equal => o.column.cmp(&token.get_src_col()),
+                    x => x,
+                });
+        let name_index = function_map.mappings[match mapping {
```
Comment (mitsuhiko) [inline src/hermes.rs]: we should use `.get()` here to prevent a panic.

File: src/hermes.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                let mut column = 0;
+
+                for mapping in line_mapping.split(',') {
+                    if mapping.is_empty() {
+                        continue;
+                    }
+
+                    let mut nums = parse_vlq_segment(mapping).ok()?.into_iter();
+
+                    column = (i64::from(column) + nums.next()?) as u32;
```
Comment (mitsuhiko) [inline src/hermes.rs]: wonder if we want to deal with the overflow explicitly here.

File: src/types.rs [hunk 3]
```diff
@@ -445,7 +451,7 @@ impl SourceMap {
     pub fn from_reader<R: Read>(rdr: R) -> Result<SourceMap> {
         match decode(rdr)? {
             DecodedMap::Regular(sm) => Ok(sm),
-            DecodedMap::Index(_) => Err(Error::IndexedSourcemap),
+            _ => Err(Error::IndexedSourcemap),
```
Comment (mitsuhiko) [inline src/types.rs]: potentially we want to change this error to `IncompatibleSourceMap`?

File: src/types.rs [hunk 4]
```diff
@@ -785,8 +791,8 @@ impl SourceMapIndex {
     /// sourcemap is encountered an error is returned.
     pub fn from_slice(slice: &[u8]) -> Result<SourceMapIndex> {
         match decode_slice(slice)? {
-            DecodedMap::Regular(_) => Err(Error::RegularSourcemap),
             DecodedMap::Index(smi) => Ok(smi),
+            _ => Err(Error::RegularSourcemap),
```
Comment (mitsuhiko) [inline src/types.rs]: Same here

File: src/hermes.rs (hunk 2 above)
Comment (Swatinem) [reply src/hermes.rs]: hm, the regular sourcemap code doesn’t either.

File: src/types.rs [hunk 5]
```diff
@@ -55,11 +56,14 @@ impl<'a> Default for RewriteOptions<'a> {
 /// Usually the two things are too distinct to provide a common
 /// interface however for token lookup and writing back into a writer
 /// general methods are provided.
+#[non_exhaustive]
```
Comment (jan-auer) [inline src/types.rs]: Is it really worth making this non_exhaustive? Introducing new subtypes might likely require users to make changes to handle the new type. Therefore, having a major bump when introducing a new variant would be most explicit.

Comment (Swatinem) [reply src/types.rs]: I agree. It also makes dealing with this type easier downstream.

File: src/hermes.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    pub fn to_writer<W: Write>(&self, w: W) -> Result<()> {
+        encode(self, w)
+    }
+
+    /// Given a bytecode offset, this will find the enclosing scopes function
+    /// name.
+    pub fn get_original_function_name(&self, bytecode_offset: u32) -> Option<&str> {
+        let token = self.sm.lookup_token(0, bytecode_offset)?;
+
+        let function_map = &self.function_maps[token.get_src_id() as usize].as_ref()?;
```
Comment (mitsuhiko) [inline src/hermes.rs]: actually one last change: just in case let's do a `.get()` here.

**Comment by HazAT** [comment]
Confirmed, just tested it end to end, it works!

**Pull request merged**